Add `path` field check in multi project repos

SDBowen opened the pull request:
Closes #183 :tada:

SDBowen commented on tasks/page/processGerbers.js:
```diff
@@ -31,15 +31,19 @@ if (require.main !== module) {
+        return (
+          info.multi[project].path === projectPath || project === projectPath
+        )
+      })
```
We have the `projectPath` here, but not the section of the yaml that contains the project's details. I am using the `projectPath` to identify the correct nested yaml section, and return that multi key (or `project` in this case).

This iterates through the `project`'s in the yaml file to check for the existence of `path`, if found, compares it with `projectPath`. If `path` doesn't exist, the `project` is compared.

The second check will become unnecessary when we depreciate using the multi keys as paths in the yaml.

The function returns the first match, which I use here:

```
kitspaceYaml = kitspaceYaml.multi[multiProjectPath]
```

kasbah replied:
Oh sorry, I missed the `[project]` bit. I think we need to switch to using more descriptive names. We have `multi[project].path`,  `project` and `projectPath` and it's getting confusing :laughing:

SDBowen replied:
Yeah, I wrote it and had to look at it for awhile!

I will try and come up with some better naming.

SDBowen commented on tasks/makeBoardsJson.js:
```diff
@@ -88,7 +88,9 @@ if (require.main !== module) {
     if (info.multi) {
       for (let project in info.multi) {
-        info.multi[project].path = project
+        if (!info.multi[project].path) {
```
I like that, will add it.

SDBowen commented on tasks/page/processBOM.js:
```diff
@@ -94,7 +96,15 @@ if (require.main !== module) {
+        kitnicYaml.multi[project].path === projectPath ||
+        project === projectPath
+      )
+    })
```
Move this out to a helper function? Would it make sense to place it in the `utils.js` file?

kasbah replied:
I guess so. It might not be worth it just on it's own.

It would be really nice to refactor all of the tasks in a way that they can later be re-used when we manage to move away from using ninja to build the site i.e. allow us to build a page dynamically when a user uploads their files. That's a much bigger task though.

SDBowen replied:
Would be willing to (slowly) work on that.

For now, I moved this logic into the `utils.js` file.

kasbah commented on tasks/page/processBOM.js:
```diff
@@ -91,22 +91,22 @@ if (require.main !== module) {
     file = fs.readFileSync(`${repoRootPath}/kitspace.yml`)
```
You missed the declaration of `kitnicYaml` above.

kasbah commented:
Looks like build is reporting as failing because `https://gogs.zenerdio.de/sebastian/satnogs-diplexer` is down.

SDBowen commented:
Does http://add-project-path-to-yaml.preview.kitspace.org/ still get updated?

I did not see the repo I added on there.

kasbah commented:
Yes, it should still be uploaded. Can't see the particle detector either though. Let's see when the latest build completes...

kasbah commented:
Build finally worked: http://add-project-path-to-yaml.preview.kitspace.org/boards/github.com/kitspace-forks/DIY_particle_detector/hardware/V1.2/

kasbah commented:
Looking good! Do you think you could throw in the `readme` path (#183). That example project would look way better if we could point both versions to the root `README.md`.

kasbah commented:
Wait, does the path do anything beside look for the readme at the moment? Are defaults `gerbers` and `1-click-bom.tsv` looked for if the fields aren't defined in the yaml?

SDBowen commented:
It is used here in the `configure` file, which later uses the folder paths in the build process.  

https://github.com/kitspace/kitspace/blob/31e073fc89ef298e3848c9338cccccbeac05155b/configure#L131-L166


I also see I used this logic in a few places which I think we can get rid of, making use of the folder path from above.

https://github.com/kitspace/kitspace/blob/2d92bfd9a1a97f77afed975759cd512f5d200a82/tasks/page/processBOM.js#L16-L22

SDBowen commented:
Testing with taking out the `bom` and `gerbers` path of a multi repo; the gerbers will be found, but it will fail if the `bom` path is missing.

https://github.com/kitspace/kitspace/blob/2d92bfd9a1a97f77afed975759cd512f5d200a82/tasks/page/processBOM.js#L45-L49

This section isn't working as intended with `multi` repos. Generates an incomplete path:

![Screenshot from 2019-12-04 22-22-56](https://user-images.githubusercontent.com/22715037/70182733-b861e780-16e4-11ea-9b43-fbf8d05a84a1.png)

kasbah commented:
It finding the gerbers may just be because it scans the whole repo for suitable files. Is there a `.tsv` file when it shows that error for the bom? Our example has `.csv` files and they are in a `bom/` folder.

SDBowen commented:
Yes, I moved and renamed the file in that case.

It seems the path isn't being built up correctly. I will look at it some
more tomorrow and better understand what's going on there.


On Wed, Dec 4, 2019, 23:29 Kaspar Emanuel <notifications@github.com> wrote:

> It finding the gerbers may just be because it scans the whole repo for
> suitable files. Is there a .tsv file when it shows that error for the
> bom? Our example has .csv files and they are in a bom/ folder.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kitspace/kitspace/pull/215?email_source=notifications&email_token=AFNJVHKFTC2SMG64WAKLOYLQXAVMRA5CNFSM4JUAVRJ2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEF6XBSA#issuecomment-561869000>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFNJVHOZ7QFSW2QIXDN5HSLQXAVMRANCNFSM4JUAVRJQ>
> .
>

kasbah commented:
Sorry to go back and forth on this but I think it actually makes much more sense to just be able to set a `readme` field rather than a `path` field. We may have to revisit the `path` field down the road but I think it's actually confusing things at the moment, because it doesn't do all that much really.

SDBowen commented:
Got a bit busy before the holidays, sorry for disappearing on this.

I agree the path field is confusing things, and I think they way I originally implemented it can be cleaned up. 

Do you think a new branch to add the `readme` field is the way to go? We can close this PR and perhaps I will revisit it once the `readme` field is added?

kasbah commented:
Sure, no problem, happy holidays! Up to you if you want to use this branch for `readme` or if you want to start a new one. 

On 22 December 2019 13:31:30 GMT, Steven Bowen <notifications@github.com> wrote:
>Got a bit busy before the holidays, sorry for disappearing on this.
>
>I agree the path field is confusing things, and I think they way I
>originally implemented it can be cleaned up. 
>
>Do you think a new branch to add the `readme` field is the way to go?
>We can close this PR and perhaps I will revisit it once the `readme`
>field is added?
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/kitspace/kitspace/pull/215#issuecomment-568262196

-- 
Sent from my mobile. Please excuse my brevity.

kasbah commented:
Closed in favor of #220
